domain/src/test/java/org/oppia/domain/profile/ProfileManagementControllerTest.kt

```diff
@@ -95,8 +122,392 @@ class ProfileManagementControllerTest {
 
   @Test
   @ExperimentalCoroutinesApi
-  fun testProfileManagementController() = runBlockingTest(coroutineContext) {
-    // TODO(#16): Finish test cases with full implementation
+  fun testAddProfile_addProfile_checkProfileIsAdded() = runBlockingTest(coroutineContext) {
+    profileManagementController.addProfile("James", "123", null, true).observeForever(mockUpdateResultObserver)
+    advanceUntilIdle()
+
+    val profileDatabase = ProfileDatabase.parseFrom(FileInputStream(File(context.filesDir, "profile_database.cache")))
+
+    verify(mockUpdateResultObserver, atLeastOnce()).onChanged(updateResultCaptor.capture())
+    assertThat(updateResultCaptor.value.isSuccess()).isTrue()
+    val profile = profileDatabase.profilesMap[0]
+    assertThat(profile?.name).isEqualTo("James")
+    assertThat(profile?.pin).isEqualTo("123")
+    assertThat(profile?.allowDownloadAccess).isEqualTo(true)
+    assertThat(profile?.id?.internalId).isEqualTo(0)
+    assertThat(File(getAbsoluteDirPath("0")).isDirectory).isTrue()
+  }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testAddProfile_addProfileWithNotUniqueName_checkResultIsFailure() = runBlockingTest(coroutineContext) {
+    addTestProfiles()
+    advanceUntilIdle()
+
+    profileManagementController.addProfile("JAMES", "321", null, true).observeForever(mockUpdateResultObserver)
+
+    verify(mockUpdateResultObserver, atLeastOnce()).onChanged(updateResultCaptor.capture())
+    assertThat(updateResultCaptor.value.isFailure()).isTrue()
```

Comment:
Here & elsewhere: we should check to make sure the correct failure is being propagated. E.g.

```kotlin
assertThat(updateResultCaptor.value.getErrorOrNull()).hasMessageThat().contains("Part of message")
```

Reply:
Updated the failure tests to check for the error message.